eos-updater-flatpak-installer/main.c

```diff
@@ -0,0 +1,316 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright © 2017 Endless Mobile, Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Sam Spilsbury <sam@endlessm.com>
+ */
+
+#include <stdlib.h>
+#include <glib.h>
+#include <locale.h>
+#include <flatpak.h>
+#include <ostree.h>
+#include <sysexits.h>
+#include <libeos-updater-util/util.h>
+
+static GPtrArray *
+parse_flatpaks_to_deploy_from_contents (const gchar *contents,
+                                        GError      **error)
+{
+  g_auto(GStrv) split_contents = g_strsplit (contents, "\n", -1);
+  g_autoptr(GPtrArray) flatpaks_to_deploy = g_ptr_array_new_full (g_strv_length (split_contents), g_object_unref);
+  GStrv split_contents_iter = split_contents;
+
+  for (; *split_contents_iter; ++split_contents_iter)
+    {
+      g_autofree gchar *remote = NULL;
+      g_autofree gchar *ref = NULL;
+      FlatpakRef *flatpak_ref = NULL;
+
+      if (!ostree_parse_refspec (*split_contents_iter, &remote, &ref, error))
+        return FALSE;
+
+      flatpak_ref = flatpak_ref_parse (ref, error);
+
+      if (!flatpak_ref)
+        return FALSE;
+
+      g_ptr_array_add (flatpaks_to_deploy,
+                       g_object_new (FLATPAK_TYPE_REMOTE_REF,
+                                     "remote-name", remote,
+                                     "name", flatpak_ref_get_name (flatpak_ref),
+                                     "kind", flatpak_ref_get_kind (flatpak_ref),
+                                     NULL));
+    }
+
+  return g_steal_pointer (&flatpaks_to_deploy);
+}
+
+static GPtrArray *
+get_flatpaks_to_deploy (GFile *to_deploy_file_path, GError **error)
+{
+  g_autofree gchar *contents = NULL;
+
+  if (!g_file_get_contents (g_file_get_path (to_deploy_file_path),
```

Comment:
The return value of `g_file_get_path()` is leaked here. Consider using `g_file_load_contents()` that takes a `GFile` directly.